### ISSUE TITLE
🐛 Bug Report: Secretredaction still not working properly

### ISSUE THREAD

**Issue by dariozachow** [backstage/backstage#24144]
### 📜 Description

With the new 1.25.0 release of backstage I was confident that the secret redaction was working, but I still have logs in templates, where Secrets are displayed as plain text. Mainly when writing to the `secrets` object in templates and writing secrets to the output, thus no longer being a secret.

### 👍 Expected behavior

No secrets should be shown in plain text in the scaffolder as every user can access the logs of each task.

### 👎 Actual Behavior with Screenshots

The first occurrence of Secrets is when there are used in the `template:fetch` step. The secret `cdaasToken` is created within a custom action which creates a GitLab Group-Accesstoken and writing it to the output of the action. Therefore there is no way of marking the output as a Secret and it's displayed in plain text.
![Screenshot 2024-04-10 at 10 57 04](https://github.com/backstage/backstage/assets/147033205/8c213831-0ae7-4b2c-bd4e-01fe2de0b0ae)

The other problem within the scaffolder is that secrets, which are set over FieldExtensions, like the RepoUrlPicker are not properly marked as secrets. Thus when using the `${{ secrets.OAUTH_TOKEN }}` the output is in plain text and not redacted. 

![Screenshot 2024-04-10 at 11 08 44](https://github.com/backstage/backstage/assets/147033205/7c618e03-eac5-485e-8a6b-e6edc25f4a24)

The token here is a reference to `secrets.OAUTH_TOKEN`, where the Component referenced as the FieldExtension is writing into the useTemplateSecrets with `setSecrets({[OAUTH_TOKEN]: token})`

### 👟 Reproduction steps

This can reproduced, when creating a simple template, which takes the token from RepoUrlPicker and writes it into a template.

And the other Problem can be reproduced by logging the output of a custom action (No suprise here because there is no way of marking outputs from custom actions as secrets).

### 📃 Provide the context for the Bug.

The problem is that everybody can read the logstream of other user with the link to the task, thus giving them the ability to read all secrets which are not redacted.

### 🖥️ Your Environment

OS:   Darwin 23.4.0 - darwin/arm64
node: v18.18.2
yarn: 3.7.0
cli:  0.26.2 (installed)
backstage:  1.25.0

Dependencies:
  @backstage/app-defaults                                          1.5.3
  @backstage/backend-app-api                                       0.5.14, 0.6.2
  @backstage/backend-common                                        0.19.10, 0.21.6
  @backstage/backend-defaults                                      0.2.16
  @backstage/backend-dev-utils                                     0.1.4
  @backstage/backend-openapi-utils                                 0.1.9
  @backstage/backend-plugin-api                                    0.6.16
  @backstage/backend-tasks                                         0.5.21
  @backstage/backend-test-utils                                    0.3.6
  @backstage/catalog-client                                        1.6.3
  @backstage/catalog-model                                         1.4.5
  @backstage/cli-common                                            0.1.13
  @backstage/cli-node                                              0.2.4
  @backstage/cli                                                   0.24.0, 0.26.2
  @backstage/config-loader                                         1.7.0
  @backstage/config                                                1.2.0
  @backstage/core-app-api                                          1.12.3
  @backstage/core-compat-api                                       0.2.3
  @backstage/core-components                                       0.13.10, 0.14.3
  @backstage/core-plugin-api                                       1.9.1
  @backstage/dev-utils                                             1.0.30
  @backstage/errors                                                1.2.4
  @backstage/eslint-plugin                                         0.1.6
  @backstage/frontend-plugin-api                                   0.6.3
  @backstage/integration-aws-node                                  0.1.12
  @backstage/integration-react                                     1.1.25
  @backstage/integration                                           1.9.1
  @backstage/plugin-api-docs                                       0.11.3
  @backstage/plugin-app-backend                                    0.3.64
  @backstage/plugin-app-node                                       0.1.16
  @backstage/plugin-auth-backend-module-atlassian-provider         0.1.8
  @backstage/plugin-auth-backend-module-aws-alb-provider           0.1.8
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.2.11
  @backstage/plugin-auth-backend-module-github-provider            0.1.13
  @backstage/plugin-auth-backend-module-gitlab-provider            0.1.13
  @backstage/plugin-auth-backend-module-google-provider            0.1.13
  @backstage/plugin-auth-backend-module-microsoft-provider         0.1.11
  @backstage/plugin-auth-backend-module-oauth2-provider            0.1.13
  @backstage/plugin-auth-backend-module-oauth2-proxy-provider      0.1.9
  @backstage/plugin-auth-backend-module-oidc-provider              0.1.7
  @backstage/plugin-auth-backend-module-okta-provider              0.0.9
  @backstage/plugin-auth-backend                                   0.22.3
  @backstage/plugin-auth-node                                      0.4.11
  @backstage/plugin-auth-react                                     0.0.3
  @backstage/plugin-catalog-backend-module-gitlab                  0.3.14
  @backstage/plugin-catalog-backend-module-openapi                 0.1.34
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.1.14
  @backstage/plugin-catalog-backend-module-unprocessed             0.4.3
  @backstage/plugin-catalog-backend                                1.21.0
  @backstage/plugin-catalog-common                                 1.0.22
  @backstage/plugin-catalog-graph                                  0.4.3
  @backstage/plugin-catalog-import                                 0.10.9
  @backstage/plugin-catalog-node                                   1.11.0
  @backstage/plugin-catalog-react                                  1.11.2
  @backstage/plugin-catalog-unprocessed-entities-common            0.0.1
  @backstage/plugin-catalog                                        1.18.2
  @backstage/plugin-events-backend-module-gitlab                   0.2.2
  @backstage/plugin-events-backend-test-utils                      0.1.26
  @backstage/plugin-events-backend                                 0.3.2
  @backstage/plugin-events-node                                    0.3.2
  @backstage/plugin-explore-common                                 0.0.2
  @backstage/plugin-github-actions                                 0.6.14
  @backstage/plugin-home-react                                     0.1.11
  @backstage/plugin-home                                           0.7.2
  @backstage/plugin-kubernetes-backend                             0.16.3
  @backstage/plugin-kubernetes-common                              0.7.5
  @backstage/plugin-kubernetes-node                                0.1.10
  @backstage/plugin-kubernetes-react                               0.3.3
  @backstage/plugin-kubernetes                                     0.11.8
  @backstage/plugin-org                                            0.6.23
  @backstage/plugin-permission-backend                             0.5.40
  @backstage/plugin-permission-common                              0.7.13
  @backstage/plugin-permission-node                                0.7.27
  @backstage/plugin-permission-react                               0.4.21
  @backstage/plugin-proxy-backend                                  0.4.14
  @backstage/plugin-scaffolder-backend-module-azure                0.1.8
  @backstage/plugin-scaffolder-backend-module-bitbucket-cloud      0.1.6
  @backstage/plugin-scaffolder-backend-module-bitbucket-server     0.1.6
  @backstage/plugin-scaffolder-backend-module-bitbucket            0.2.6
  @backstage/plugin-scaffolder-backend-module-gerrit               0.1.8
  @backstage/plugin-scaffolder-backend-module-gitea                0.1.6
  @backstage/plugin-scaffolder-backend-module-github               0.2.6
  @backstage/plugin-scaffolder-backend-module-gitlab               0.3.2
  @backstage/plugin-scaffolder-backend                             1.22.3
  @backstage/plugin-scaffolder-common                              1.5.1
  @backstage/plugin-scaffolder-node-test-utils                     0.1.2
  @backstage/plugin-scaffolder-node                                0.4.2
  @backstage/plugin-scaffolder-react                               1.8.3
  @backstage/plugin-scaffolder                                     1.19.2
  @backstage/plugin-search-backend-module-catalog                  0.1.21
  @backstage/plugin-search-backend-module-explore                  0.1.20
  @backstage/plugin-search-backend-module-pg                       0.5.25
  @backstage/plugin-search-backend-module-techdocs                 0.1.21
  @backstage/plugin-search-backend-node                            1.2.20
  @backstage/plugin-search-backend                                 1.5.6
  @backstage/plugin-search-common                                  1.2.11
  @backstage/plugin-search-react                                   1.7.9
  @backstage/plugin-search                                         1.4.9
  @backstage/plugin-techdocs-backend                               1.10.3
  @backstage/plugin-techdocs-module-addons-contrib                 1.1.8
  @backstage/plugin-techdocs-node                                  1.12.2
  @backstage/plugin-techdocs-react                                 1.2.2
  @backstage/plugin-techdocs                                       1.10.3
  @backstage/plugin-user-settings                                  0.8.4
  @backstage/release-manifests                                     0.0.11
  @backstage/test-utils                                            1.5.3
  @backstage/theme                                                 0.4.4, 0.5.2
  @backstage/types                                                 1.1.1
  @backstage/version-bridge                                        1.0.7

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, I don't have time to work on this right now

**Comment by dariozachow** [comment]
For reference: https://github.com/backstage/backstage/issues/23413#event-12130726584 and https://github.com/backstage/backstage/pull/23453

**Comment by benjdlambert** [comment]
@dariozachow unfortunately, this isn't supported. Secrets are only allowed to be set once, and they're created on job creation, so they're not going to be passed to the logger for redaction as they're the only secrets that are used.

I guess that in the action you're doing `ctx.secrets.mySecret = 'blob'` or something right? 

We could look to implement something like `ctx.setSecret` which could be useful perhaps, especially for other use cases too like one time retrieval for outputs or something.

**Comment by dariozachow** [comment]
I think this would be helpful, but we also need this for `useTemplateSecrets` as the scaffolder does not redact those secrets aswell

**Comment by benjdlambert** [comment]
@dariozachow it should redact all secrets that are send at the initial creation? Are you not seeing that? Those are the ones that are sent with `useTemplateSecrets`?

**Comment by dariozachow** [comment]
For Reference the template:

```yaml
apiVersion: scaffolder.backstage.io/v1beta3
kind: Template
metadata:
  name: dario-test 
  title: Dario Test
  description: test
spec:
  owner: user:dariozachow
  type: service

  # These parameters are used to generate the input form in the frontend, and are
  # used to gather input data for the execution of the template.
  parameters:
    - title: user information
      required:
        - user
      properties:
        user:
          title: user
          type: string
          description: tenant to be created
    - title: Gitlab Repo Picker 
      required:
        - repoUrl
      properties:
        repoUrl:
          title: test
          type: string
          description: Pick a GitLabRepository 
          ui:field: GitLabRepoPicker
          ui:options:
            host: <my-host>
  steps:
    - id: fetch
      name: Templating catalog-info
      action: fetch:template
      input:
        url: ./
        token: ${{ secrets.USER_OAUTH_TOKEN }}
        values:
          token: ${{ secrets.USER_OAUTH_TOKEN }}
    - id: debug 
      name: debug
      action: debug:log
      input:
        message: ${{ secrets.USER_OAUTH_TOKEN }}
```

![image](https://github.com/backstage/backstage/assets/147033205/dd763b77-c6f5-4f73-94fd-4a11458b2591)

**Comment by benjdlambert** [comment]
@dariozachow gonna close this as fixed, should be in the latest `-next` release that went out yesterday 🤞

**Comment by floric** [comment]
Hey @benjdlambert , we tried the new release 1.27.2 with various secrets, safed in the `secrets`-scope with `fetch:template` and `debug:log`. Unfortunately, in comparison to our latest used release 1.26.2 now all secrets are logged. None of them are redacted. Before, at least some were marked as redacted.

Apart from the upgrade to 1.27 we didn't change anything. So unfortunately something must have been broken. I also looked into the changes in #24478  but couldn't find any problem so far.
